lib/identity_cache/query_api.rb

```diff
@@ -9,6 +9,56 @@ def prefetch_associations(includes, records)
         Cached::Prefetcher.prefetch(self, includes, records)
       end
 
+      # Get only the columns whose values are needed to manually expire caches
+      # after updating or deleting rows without triggering after_commit callbacks.
+      #
+      # 1. Pass the returned columns into Active Record's `select` or `pluck` query
+      #    method on the scope that will be used to modify the database in order to
+      #    query original for these rows that will be modified.
+      # 2. Update or delete the rows
+      # 3. Use {expire_cache_for_update} or {expire_cache_for_delete} to expires the
+      #    caches, passing in the values from the query in step 1 as the indexed_values.
+      #
+      # @return [Array<Symbol>] the array of column names
+      def cache_indexed_columns
+        @cache_indexed_columns ||= begin
+          check_for_unsupported_parent_expiration_entries
+          columns = Set.new
+          columns << primary_key.to_sym if primary_cache_index_enabled
+          cache_indexes.each do |cached_attribute|
+            columns.merge(cached_attribute.key_fields)
+          end
+          columns.to_a.freeze
+        end
+      end
+
+      def expire_cache_for_update(old_indexed_values, changes)
+        if primary_cache_index_enabled
+          id = old_indexed_values.fetch(primary_key.to_sym)
+          expire_primary_key_cache_index(id)
+        end
+        cache_indexes.each do |cached_attribute|
+          cached_attribute.expire_for_update(old_indexed_values, changes)
+        end
+        parent_expiration_entries.each do ||
```

Comment:
Missing `end` for this block?

Reply:
This line was meant to be deleted.  I'll reiterate that

> Note that I haven't yet tested or fully documented this code

I opened it early as a draft PR, since initially I was just trying to review and make recommendations for https://github.com/Shopify/identity_cache/pull/495 then decided code would better communicate this... but I didn't want to delay that communication in order to get this PR working.

I've pushed a fix to avoid breaking the existing tests... but it still needs tests for the new code paths.